contributors/design-proposals/bootstrap-discovery.md

```diff
@@ -173,6 +172,47 @@ In addition, `jws-kubeconfig-<token-id>-hash` will be set to the MD5 hash of the
+* `--token-ttl` If set, this sets the TTL for the lifetime of this token.  Defaults to 0 which means "forever"
+
+### `kubeadm join` flags
+
+* `--token` This sets the token for both discovery and bootstrap auth.
+* `--discovery-url` If set this will grab the cluster-info data (a kubeconfig) from a URL. Due to the sensitive nature of this data, we will only support https URLs.  This also supports `username:password@host` syntax for doing HTTP auth.
+* `--discovery-file` If set, this will load the cluster-info from a file.
+* `--discovery-token` If set, (or set via `--token`) then we will be using the token scheme described above.
+
+Only one of `--discovery-url`, `--discovery-file` or `--discovery-token` can be set.  If more than one is set then an error is surfaced and `kubeadm join` exits.  Setting `--token` counts as setting `--discovery-token`.
+
+### `kubeadm token` commands
+
+`kubeadm` provides a set of utilities for manipulating token secrets in a running server.
+
+* `kubeadm token create` Creates a token server side.  With no options this'll create a token that is used for discovery and TLS bootstrap.
+  * `--token` The actual token value (in `id.secret` form) to write in.  If unset, a random value is generated.
+  * `--usages` A list of usages.  Defaults to `signing,authentication`.  Signing is used to sign the config-map used for discovery.  Authentication is used to authenticate for TLS bootstrap.
+  * `--ttl` The TTL for this token.  This sets the expiration of the token as a duration from the current time.
+* `kubeadm token delete <token-id>|<token-id>.<token-secret>`
+  * Users can either just specify the id or the secret.  This will delete the token if it exists.
+* `kubeadm token list`
+  * List tokens in a table form listing out the `token-id.token-secret`, the TTL, the absolute expiration time and the usages.
+  * **Question** Support a `--json` or `-o json` way to make this info programmatic? We don't want to recreate `kubectl` here and these aren't plain API objects so we can't reuse that plumbing easily.
+* `kubeadm token generate` This currently exists but is documented here for completeness.  This pure client side method just generated a random token in the correct form.
+
+## Implementation Details
+
+After some back and forth, we decided to keep the separator in the token between the ID and Secret be a `.`.  During the 1.6 cycle, at one point `:` was implemented but then reverted.
+
+See https://github.com/kubernetes/client-go/issues/114 for details on creating a shared package with common constants for this scheme.
+
+This proposal assumes RBAC to lock things down in a couple of ways.  First, it will open up `kube-public` so that it is readable by unauthenticated users.  Next, it will make it so that the identities in the `system:bootstrappers` group can only be used to submit certs API to submit CSRs.  After a TLS certificate is created, that identity should be used instead of the bootstrap token.
```

Comment:
Initially, we're only opening `get configmap` in `kube-public` to avoid getting opening ourselves to the world too much